File: agent/src/nexus/server/routes/projects.py
"""Routes for project management: CRUD + session assignment.

Projects are named workspaces that group related chat sessions with
project-scoped instructions and an auto-created vault subfolder.
"""

from __future__ import annotations

import asyncio
import logging
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, status

from ..project_store import ProjectStore

log = logging.getLogger(__name__)

router = APIRouter()

_DB_PATH = Path("~/.nexus/sessions.sqlite").expanduser()


def _get_store() -> ProjectStore:
    return ProjectStore(_DB_PATH)


@router.get("/projects")
async def list_projects(
    limit: int = 50,
) -> list[dict]:
    store = _get_store()
    summaries = await asyncio.to_thread(store.list, limit=max(1, min(limit, 200)))
    return [
        {
            "id": s.id,
            "name": s.name,
            "description": s.description,
            "color": s.color,
            "icon": s.icon,
            "session_count": s.session_count,
            "created_at": s.created_at,
            "updated_at": s.updated_at,
        }
        for s in summaries
    ]


@router.post("/projects", status_code=status.HTTP_201_CREATED)
async def create_project(
    body: dict,
) -> dict:
    name = (body.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=422, detail="name is required")
    store = _get_store()
    project = await asyncio.to_thread(
        store.create,
        name=name,
        description=body.get("description", ""),
        instructions=body.get("instructions", ""),
        color=body.get("color", ""),
        icon=body.get("icon", ""),
    )
    return _project_to_dict(project)


@router.get("/projects/{project_id}")
async def get_project(
    project_id: str,
) -> dict:
    store = _get_store()
    project = await asyncio.to_thread(store.get, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return _project_to_dict(project)


@router.patch("/projects/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def update_project(
    project_id: str,
    body: dict,
) -> None:
    store = _get_store()
    project = await asyncio.to_thread(store.get, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    await asyncio.to_thread(store.update, project_id, **body)


@router.delete("/projects/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: str,
) -> None:
    store = _get_store()
    project = await asyncio.to_thread(store.get, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    await asyncio.to_thread(store.delete, project_id)


@router.post(
    "/projects/{project_id}/sessions/{session_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def move_session_to_project(
    project_id: str,
    session_id: str,
) -> None:
    store = _get_store()
    project = await asyncio.to_thread(store.get, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    await asyncio.to_thread(store.move_session, session_id, project_id)


@router.delete(
    "/projects/{project_id}/sessions/{session_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def remove_session_from_project(
    project_id: str,
    session_id: str,
) -> None:
    store = _get_store()
    await asyncio.to_thread(store.move_session, session_id, None)


def _project_to_dict(project: Any) -> dict:
    return {
        "id": project.id,
        "name": project.name,
        "description": project.description,
        "instructions": project.instructions,
        "vault_path": project.vault_path,
        "color": project.color,
        "icon": project.icon,
        "created_at": project.created_at,
        "updated_at": project.updated_at,
    }

File: ui/src/api/projects.ts
import { BASE } from "./base";

export interface Project {
  id: string;
  name: string;
  description: string;
  instructions: string;
  vault_path: string;
  color: string;
  icon: string;
  created_at: number;
  updated_at: number;
}

export interface ProjectSummary {
  id: string;
  name: string;
  description: string;
  color: string;
  icon: string;
  session_count: number;
  created_at: number;
  updated_at: number;
}

export async function listProjects(limit = 50): Promise<ProjectSummary[]> {
  const res = await fetch(`${BASE}/projects?limit=${limit}`);
  if (!res.ok) throw new Error("Failed to list projects");
  return res.json();
}

export async function createProject(data: {
  name: string;
  description?: string;
  instructions?: string;
  color?: string;
  icon?: string;
}): Promise<Project> {
  const res = await fetch(`${BASE}/projects`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const detail = (await res.json().catch(() => ({})))?.detail || "Failed to create project";
    throw new Error(detail);
  }
  return res.json();
}

export async function getProject(id: string): Promise<Project> {
  const res = await fetch(`${BASE}/projects/${encodeURIComponent(id)}`);
  if (!res.ok) throw new Error("Project not found");
  return res.json();
}

export async function updateProject(
  id: string,
  data: Partial<Pick<Project, "name" | "description" | "instructions" | "color" | "icon">>,
): Promise<void> {
  const res = await fetch(`${BASE}/projects/${encodeURIComponent(id)}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error("Failed to update project");
}

export async function deleteProject(id: string): Promise<void> {
  const res = await fetch(`${BASE}/projects/${encodeURIComponent(id)}`, {
    method: "DELETE",
  });
  if (!res.ok) throw new Error("Failed to delete project");
}

export async function moveSessionToProject(
  projectId: string,
  sessionId: string,
): Promise<void> {
  const res = await fetch(
    `${BASE}/projects/${encodeURIComponent(projectId)}/sessions/${encodeURIComponent(sessionId)}`,
    { method: "POST" },
  );
  if (!res.ok) throw new Error("Failed to move session");
}

export async function removeSessionFromProject(
  projectId: string,
  sessionId: string,
): Promise<void> {
  const res = await fetch(
    `${BASE}/projects/${encodeURIComponent(projectId)}/sessions/${encodeURIComponent(sessionId)}`,
    { method: "DELETE" },
  );
  if (!res.ok) throw new Error("Failed to remove session from project");
}
